lib/livebook/notebook/explore/kino/pong.livemd

```diff
@@ -1,492 +1,899 @@
-start implementing Pong! To make the objective more clear, we will
-start by visualizing the scene. To do so, we will use SVG images
-and for clarity let's build a tiny abstraction for painting those
-images.
+```elixir
+Scene.new(80, 10)
+|> Scene.add(Scene.Rect.new(0, 0, 80, 10, fill: "#d4f1f477"))
+|> Scene.add(Scene.Text.new(40, 4, "Hello world!", font_size: 3))
+|> Scene.to_svg()
+|> Kino.Image.new(:svg)
+```
+
+Perfect.
+
+## Animating scene objects
+
+So far we are able to paint static objects in the scene.
+However, all games have dynamic objects that move on the scene
+as time passes. In our case, the dynamic objects are the paddles
+and the ball.
+
+For such, we will define an `Animation` protocol with a single
+function called `step`. It receives the object, the scene, and
+is responsible for animating the object one step at a time.
 
 ```elixir
-defmodule Canvas do
-  @moduledoc """
-  A canvas for graphing geometric shapes on.
-  """
+defprotocol Animation do
+  def step(object, scene)
+end
+```
 
-  defstruct [:width, :height, :scale, :objects]
+Now let's define our first game object, the `Game.Ball` struct.
+The ball will have a radius `r`, coordinates `x` and `y`, velocities
+`dx` and `dy`, and implement both `Scene.Renderer` and `Animation`
+protocols.
 
-  def new(width, height, opts \\ []) do
-    scale = opts[:scale] || 10
-    %__MODULE__{width: width, height: height, scale: scale, objects: []}
-  end
+```elixir
+defmodule Game.Ball do
+  defstruct [:r, :x, :y, :dx, :dy]
 
-  def add_rect(canvas, x, y, width, height, opts \\ []) do
-    fill = opts[:fill] || "black"
-    object = {:rect, x, y, width, height, fill}
-    %{canvas | objects: canvas.objects ++ [object]}
+  def new(r, x, y, opts \\ []) do
+    struct!(%Game.Ball{r: r, x: x, y: y}, opts)
   end
 
-  def add_circle(canvas, x, y, r, opts \\ []) do
-    fill = opts[:fill] || "black"
-    object = {:circle, x, y, r, fill}
-    %{canvas | objects: canvas.objects ++ [object]}
+  defimpl Scene.Renderer do
+    def to_svg(%Game.Ball{x: x, y: y, r: r}) do
+      """
+      <circle cx="#{x}" cy="#{y}" r="#{r}" fill="tomato" />
+      """
+    end
   end
 
-  def add_text(canvas, x, y, text, opts \\ []) do
-    fill = opts[:fill] || "black"
-    font_size = opts[:font_size] || 1
-    object = {:text, x, y, text, font_size, fill}
-    %{canvas | objects: canvas.objects ++ [object]}
+  defimpl Animation do
+    # On each step, increase x by dx, and y by dy.
+    def step(ball, _scene) do
+      %{ball | x: ball.x + ball.dx, y: ball.y + ball.dy}
+    end
   end
+end
+```
+
+We can now paint the ball in the scene:
+
+```elixir
+Scene.new(80, 10)
+|> Scene.add(Game.Ball.new(2, 40, 4))
+|> Scene.to_svg()
+|> Kino.Image.new(:svg)
+```
+
+However, we can also animate the ball over the scene using `Kino.Frame`.
+Let's define the frame, the scene, and the ball:
+
+```elixir
+# The frame we will render scene images on top of
+frame = Kino.Frame.new() |> Kino.render()
```

Comment:
Maybe create the frame last, or explicitly return it from the cell, so the ball doesn't get inspected?